File: UITableViewThreeEventsDemo/Demo-objc/Demo-objc/CellSuperViewController.m
//
//  CellSuperViewController.m
//  Demo-20160419
//
//  Created by crw on 4/19/16.
//  Copyright © 2016 crw. All rights reserved.
//

#import "CellSuperViewController.h"
#import "Cell.h"

@interface CellSuperViewController()

@property (nonatomic, strong) NSMutableArray     *dataArray;

@end

@implementation CellSuperViewController

#pragma mark - life cycle
- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.dataArray = [Model getData];
    
    [self.tableView reloadData];
}

#pragma mark - UITableViewDelegate
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    
    static NSString *identifier = @"Cell";
    
    Cell *cell = [tableView dequeueReusableCellWithIdentifier:identifier forIndexPath:indexPath];
    
    cell.model = self.dataArray[indexPath.row];
    
    [cell.delBtn addTarget:self action:@selector(delItem:) forControlEvents:UIControlEventTouchUpInside];
    
    return cell;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    return self.dataArray.count;
}

#pragma mark - Events
- (void)delItem:(UIButton *)btn{
    
    UIView *view = btn;
    while (view != nil && ![view isKindOfClass:[UITableViewCell class]]) {
        view = [view superview];
    }
    
    Cell *cell = (Cell *)view;
    NSIndexPath *indexPath = [self.tableView indexPathForCell:cell];
    
    [self.dataArray removeObjectAtIndex:indexPath.row];
    [self.tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationMiddle];
    
}

@end

File: UITableViewThreeEventsDemo/Demo-swift/Demo-swift/CellSuperViewController.swift
//
//  CellSuperViewController.swift
//  Demo-20160418
//
//  Created by crw on 4/20/16.
//  Copyright © 2016 crw. All rights reserved.
//

import UIKit

class CellSuperViewController: UITableViewController {
    
    var dataArray = [Model]()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        dataArray = Model.getData()
        
        tableView.reloadData()
    }
    
}

extension CellSuperViewController
{
    
    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return dataArray.count
    }
    
    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCellWithIdentifier(ID, forIndexPath: indexPath) as! Cell
        cell.model = dataArray[indexPath.row]
        cell.delBtn.addTarget(self, action:#selector(deleteItem(_:)), forControlEvents: UIControlEvents.TouchUpInside)
        return cell
    }
    
    func deleteItem(btn : UIButton) {
        
        var view : UIView = btn;
        
        while (!view.isKindOfClass(UITableViewCell.self)) {
            view = view.superview!;
        }
        
        let cell = view as! Cell;
        
        if let indexPath = tableView.indexPathForCell(cell) {
            dataArray.removeAtIndex(indexPath.row)
            tableView.deleteRowsAtIndexPaths([indexPath], withRowAnimation: UITableViewRowAnimation.Fade)
        }
    }
}